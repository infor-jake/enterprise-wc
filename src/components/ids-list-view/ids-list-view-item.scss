:host {
  .ids-list-view-item {
    border: none;
    margin: 0;
    padding: 0;
  }
}
