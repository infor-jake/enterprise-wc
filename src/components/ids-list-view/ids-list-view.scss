:host([selectable]) {
  .ids-list-view .ids-list-view-body {
    div[part='list-item'],
    li {
      cursor: var(--ids-cursor-pointer);
    }
  }
}

:host([no-margins]) {
  .ids-list-view .ids-list-view-body {
    div[part='list-item'],
    li {
      &:first-child {
        margin-block-start: var(--ids-space-none);
      }
    }
  }
}

.ids-list-view {
  height: var(--ids-list-view-height);
  overflow: auto;
  scroll-behavior: smooth;
  position: relative;

  .ids-list-view-body {
    margin: 0;
    padding: 0;

    &:focus {
      border-color: var(--ids-list-view-color-border-focus);
    }

    div[part='list-item'],
    li {
      background-color: var(--ids-list-view-item-color-background-default);
      border: var(--ids-list-view-item-border);
      border-top-color: var(--ids-list-view-item-color-border-top);
      contain: content;
      list-style-type: none;
      padding: var(--ids-list-view-item-padding);

      &:first-child {
        margin-block-start: var(--ids-space-20);
      }

      &:focus {
        border-color: var(--ids-list-view-item-color-border-focus);
        box-shadow: var(--ids-list-view-item-shadow-focus);
        outline: var(--ids-list-view-item-outline-focus);

        + div[part='list-item'] {
          border-top-color: transparent;
        }
      }

      &.sortable {
        background-color: transparent;
        border: none;
        cursor: inherit;

        &:first-child {
          margin-block-start: var(--ids-space-none);
        }
      }

      &[disabled] {
        color: var(--ids-list-view-item-color-text-disabled);
        cursor: var(--ids-cursor-default);
        pointer-events: none;
        user-select: none;

        &:focus {
          border-color: var(--ids-list-view-item-color-border-disabled-focus);
          border-top-color: var(--ids-list-view-item-top-color-border-disabled-focus);
          box-shadow: var(--ids-list-view-item-shadow-disabled-focus);
        }

        ids-badge::part(badge) {
          background-color: var(--ids-list-view-item-color-background-disabled);
          border-color: var(--ids-list-view-item-color-border-disabled);
          color: var(--ids-list-view-item-badge-color-text-disabled);
        }

        ids-hyperlink::part(link) {
          color: var(--ids-list-view-item-color-text-link-disabled);
        }

        ids-image {
          opacity: var(--ids-list-view-item-opacity-disabled);
        }
      }
    }
  }

  &.has-searchfield .ids-list-view-body {
    div[part='list-item'],
    li {
      &:first-child {
        margin-block-start: 8px;
      }
    }
  }

  &.selectable-single .ids-list-view-body,
  &.selectable-multiple .ids-list-view-body,
  &.selectable-mixed .ids-list-view-body {
    div[part='list-item']:not([disabled]):not(.sortable),
    li:not([disabled]):not(.sortable) {
      &:hover,
      &[activated],
      &[selected]:not([hide-selected-color]) {
        background-color: var(--ids-list-view-color-background-selected);
      }
    }
  }

  &.selectable-multiple,
  &.selectable-mixed {
    .list-item-area {
      display: flex;
      align-items: center;

      .list-item-checkbox {
        padding-inline-end: var(--ids-space-70);
        flex-grow: 0;
        flex-shrink: 0;
        max-height: var(--ids-space-60);
        max-width: var(--ids-space-60);
      }

      .list-item-checkbox[disabled]::part(checkmark) {
        background-color: transparent;
      }
    }
  }

  &.is-scrolling {
    box-shadow: var(--ids-scroll-shadow);
    border-top: var(--ids-scroll-color-border-top);
  }
}

ids-swappable-item {
  border: 1px solid transparent;
  border-bottom-color: var(--ids-color-neutral-20);
}

ids-swappable-item[over] {
  border: 1px dashed var(--ids-color-neutral-80);
}

.ids-list-view-search {
  margin-bottom: 8px;
}
