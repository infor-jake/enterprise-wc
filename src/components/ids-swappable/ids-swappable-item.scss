:host {
  font-family: var(--ids-swappable-item-font-family);
  border: var(--ids-swappable-item-border-default);
  cursor: var(--ids-swappable-item-cursor-default);
  display: block;
  transition: height 0.4s ease, padding 0.4s ease;
}

:host(:hover) {
  background-color: var(--ids-swappable-item-color-background-hover);
  position: relative;

  &::before {
    position: absolute;
    left: 0;
    top: 50%;
    transform: translate(5px, -47%);
    font-size: var(--ids-swappable-item-font-size-hover);
    content: var(--ids-swappable-item-content-hover);
  }
}

:host([dir='rtl']:hover) {
  &::before {
    left: unset;
    right: 5px;
    transform: translate(-1px, -47%);
  }
}

:host(:focus) {
  border: var(--ids-swappable-item-border-focus);
  box-shadow: var(--ids-swappable-item-shadow-focus);
  outline: var(--ids-swappable-item-outline-focus);
}

:host([selected]) {
  cursor: var(--ids-swappable-item-cursor-selected);
}

:host([dragging]) {
  opacity: var(--ids-swappable-item-opacity-dragging);
}

:host([dragging]:not(.slotted-ids-tab)),
:host([selected]:not(.slotted-ids-tab)) {
  background-color: var(--ids-swappable-item-color-background-selected);
  color: var(--ids-swappable-item-color-text-selected);
}

:host([over]) {
  border: var(--ids-swappable-item-over-border);
}

:host(.p-8) {
  padding-block: var(--ids-space-40);
  padding-inline-end: var(--ids-space-40);
  padding-inline-start: var(--ids-space-60);
}

:host(.is-hidden) {
  height: var(--ids-space-none);
  padding: var(--ids-space-none);
  border: none;
  overflow: hidden;
}

// Fixes slotted IdsTabs to appear flush with their containers
// (IdsSwappableItem causes a 1px boundary as a result of its required transparent border)
// @TODO: Relocate this to another component / review wrapping, separation of concerns
::slotted(ids-tab) {
  position: relative;
  margin: -1px;
}

:host(.slotted-ids-tab:not(.orientation-vertical)) {
  margin-inline-start: var(--ids-space-30);
}
