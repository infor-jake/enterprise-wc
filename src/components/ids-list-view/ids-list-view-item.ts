import { customElement, scss } from '../../core/ids-decorators';
import { attributes } from '../../core/ids-attributes';
import { stringToBool } from '../../utils/ids-string-utils/ids-string-utils';
import IdsEventsMixin from '../../mixins/ids-events-mixin/ids-events-mixin';
import IdsElement from '../../core/ids-element';
import '../ids-checkbox/ids-checkbox';
import '../ids-swappable/ids-swappable';
import '../ids-swappable/ids-swappable-item';
import styles from './ids-list-view-item.scss';

const Base = IdsEventsMixin(
  IdsElement
);

/**
 * IDS List View Item Component
 * @type {IdsListViewItem}
 * @inherits IdsElement
 */
@customElement('ids-list-view-item')
@scss(styles)
export default class IdsListViewItem extends Base {
  #parentListView: HTMLElement | null = null;

  /**
   * Get the parent IdsListView that contains this list-view-item
   * @returns {HTMLElement | null} this list-item's parent slot
   */
  get parentListView(): HTMLElement | null {
    return this.#parentListView;
  }

  /**
   * Get the top-most wrapper element that contains this list-view-item
   * @returns {HTMLElement | null} this list-item's parent slot
   */
  get parentListItemWrapper(): HTMLElement | null {
    return this.parentListView?.shadowRoot?.querySelector(`[index="${this.rowIndex}"]`) ?? null;
  }

  /**
   * Get slot that was autogenerated in IdsListView to contain this list-view-item
   * @returns {HTMLSlotElement | null} this list-item's parent slot
   */
  get parentSlot(): HTMLSlotElement | null {
    return this.parentListView?.shadowRoot?.querySelector<HTMLSlotElement>(`slot[name="${this.slot}"]`) ?? null;
  }

  /**
   * Invoked each time the custom element is appended into a document-connected element.
   */
  connectedCallback() {
    super.connectedCallback();
    this.#parentListView = this.parentElement;
  }

  /**
   * Invoked each time the custom element is removed from a document-connected element.
   */
  disconnectedCallback() {
    if (document.body.contains(this.parentListView)) {
      super.disconnectedCallback();
      this.removeAttribute?.('slot');
      // this.parentListItemWrapper?.remove?.();
      (this.parentListView as any)?.disconnectedCallback?.();
    }
  }

  /**
   * Inner template contents
   * @returns {string} The template
   */
  template(): string {
    return `<slot></slot>`;
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array} The attributes in an array
   */
  static get attributes() {
    return [
      ...super.attributes,
      attributes.ACTIVE,
      attributes.DISABLED,
      attributes.SELECTED,
      attributes.ROW_INDEX,
    ];
  }

  /**
   * Set the row index. This index can be used to lazy-load data from IdsListView.data.
   * @param {number} value the index
   */
  set rowIndex(value: number) {
    if (value !== null && value >= 0) {
      this.setAttribute(attributes.ROW_INDEX, String(value));
    } else {
      this.removeAttribute(attributes.ROW_INDEX);
    }
  }

  /**
   * Get the row index. This index can be used to lazy-load data from IdsListView.data.
   * @returns {number} this list-view-item's index in parent IdsListView
   */
  get rowIndex(): number { return Number(this.getAttribute(attributes.ROW_INDEX) ?? -1); }

  /**
   * Wrapper function that adds interface to match dataset interface.
   * @returns {boolean} true/false
   */
  get itemActivated(): boolean { return this.active; }

  /**
   * Get the list-item active state.
   * @returns {boolean} true/false
   */
  get active(): boolean { return this.hasAttribute(attributes.ACTIVE); }

  /**
   * Set the list-item active state.
   * @param {boolean} value true/false
   */
  set active(value: boolean) {
    if (stringToBool(value)) {
      this.setAttribute(attributes.ACTIVE, 'true');
    } else {
      this.removeAttribute(attributes.ACTIVE);
    }
  }

  /**
   * Set the list-item disabled state.
   * @param {boolean} value true/false
   */
  set disabled(value: boolean) {
    if (stringToBool(value)) {
      this.setAttribute(attributes.DISABLED, 'true');
    } else {
      this.removeAttribute(attributes.DISABLED);
    }
  }

  /**
   * Get the list-item disabled state.
   * @returns {boolean} true/false
   */
  get disabled(): boolean { return this.hasAttribute(attributes.DISABLED); }

  /**
   * Wrapper function that adds interface to match dataset interface.
   * @returns {boolean} true/false
   */
  get itemSelected(): boolean { return this.selected; }

  /**
   * Get the list-item selected state.
   * @returns {boolean} true/false
   */
  get selected(): boolean { return this.hasAttribute(attributes.SELECTED); }

  /**
   * Set the list-item selected state.
   * @param {boolean} value true/false
   */
  set selected(value: boolean) {
    if (stringToBool(value)) {
      this.setAttribute(attributes.SELECTED, 'true');
    } else {
      this.removeAttribute(attributes.SELECTED);
    }
  }
}
