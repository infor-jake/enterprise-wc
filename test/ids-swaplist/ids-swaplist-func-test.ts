/**
 * @jest-environment jsdom
 */
import IdsSwapList from '../../src/components/ids-swaplist/ids-swaplist';
import '../../src/components/ids-swappable/ids-swappable-item';
import '../../src/components/ids-list-view/ids-list-view';
import IdsSwappable from '../../src/components/ids-swappable/ids-swappable';
import dataset from '../../src/assets/data/swaplist-data.json';

const HTMLSnippets = {
  SWAPLIST_COMPONENT: (
    `<ids-swaplist id="swaplist-1"></ids-swaplist>`
  ),
};

describe('IdsSwapList Component', () => {
  let idsSwapList: any;
  let idsSwappable: any;

  const createElemViaTemplate = async (innerHTML: any) => {
    idsSwapList?.remove?.();

    const template = document.createElement('template');
    template.innerHTML = innerHTML;
    idsSwapList = template.content.childNodes[0];

    document.body.appendChild(idsSwapList);

    return idsSwapList;
  };

  beforeEach(async () => {
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb: any) => cb());

    idsSwappable = new IdsSwappable();
    idsSwapList = new IdsSwapList();
    idsSwapList.innerHTML = '<ids-text>${city}</ids-text>'; //eslint-disable-line

    document.body.appendChild(idsSwapList);
    idsSwapList.data = dataset;
  });

  afterEach(async () => {
    idsSwapList?.remove();
  });

  it('renders with no errors', async () => {
    idsSwapList = await createElemViaTemplate(HTMLSnippets.SWAPLIST_COMPONENT);

    const errors = jest.spyOn(global.console, 'error');
    expect(document.querySelectorAll('ids-swaplist').length).toEqual(1);
    idsSwapList.innerHTML = '<template><ids-text>${city}</ids-text></template>'; //eslint-disable-line
    idsSwapList.data = dataset;

    idsSwapList.remove();

    idsSwapList = await createElemViaTemplate(HTMLSnippets.SWAPLIST_COMPONENT);
    expect(document.querySelectorAll('ids-swaplist').length).toEqual(1);

    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', async () => {
    idsSwapList = await createElemViaTemplate(HTMLSnippets.SWAPLIST_COMPONENT);
    idsSwapList.template();
    idsSwapList.data = dataset;
    expect(idsSwapList.outerHTML).toMatchSnapshot();
  });

  it('can swap item to next/previous list on click event', async () => {
    const event0 = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window
    });
    idsSwapList.offEvent('click', idsSwapList);
    idsSwapList.container.dispatchEvent(event0);

    const listItem = idsSwapList.container.querySelector('ids-swappable-item');
    const event = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window
    });
    listItem.offEvent('click', listItem);
    listItem.dispatchEvent(event);

    const leftArrowBtn = idsSwapList.shadowRoot.querySelector('.left-arrow');
    const event2 = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window
    });
    leftArrowBtn.offEvent('click', leftArrowBtn);
    leftArrowBtn.dispatchEvent(event2);

    const rightArrowBtn = idsSwapList.shadowRoot.querySelector('.right-arrow');
    const event3 = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window
    });
    rightArrowBtn.offEvent('click', rightArrowBtn);
    rightArrowBtn.dispatchEvent(event3);

    const currentCard1 = leftArrowBtn.parentElement.parentElement.parentElement;
    const currentList1 = currentCard1.querySelector('ids-swappable');
    const currentCard2 = rightArrowBtn.parentElement.parentElement.parentElement;
    const currentList2 = currentCard2.querySelector('ids-swappable');

    setTimeout(() => {
      expect(listItem.selected).toBe('selected');
      expect(idsSwappable.length).toBe(1);
      expect(currentList1.length).toBe(1);
      expect(currentList2.length).toBe(1);
    }, 20);
  });

  it('can swap item to next/previous list on touch event', async () => {
    const event0 = new TouchEvent('touchend', {
      touches: [{
        identifier: 123,
        target: idsSwapList.container,
        pageX: 0,
        pageY: 0,
        clientX: 0,
        clientY: 0,
        force: 0,
        radiusX: 0,
        radiusY: 0,
        rotationAngle: 0,
        screenX: 0,
        screenY: 0
      }],
      bubbles: true,
      cancelable: true,
      view: window
    });
    idsSwapList.container.dispatchEvent(event0);

    const listItem = idsSwapList.container.querySelector('ids-swappable-item');
    const event = new TouchEvent('touchend', {
      touches: [{
        identifier: 123,
        pageX: 0,
        pageY: 0,
        target: listItem,
        clientX: 0,
        clientY: 0,
        force: 0,
        radiusX: 0,
        radiusY: 0,
        rotationAngle: 0,
        screenX: 0,
        screenY: 0
      }],
      bubbles: true,
      cancelable: true,
      view: window
    });
    listItem.offEvent('touchend', listItem);
    listItem.dispatchEvent(event);

    const leftArrowBtn = idsSwapList.shadowRoot.querySelector('.left-arrow');
    const event2 = new TouchEvent('touchend', {
      touches: [{
        identifier: 123,
        pageX: 0,
        pageY: 0,
        target: leftArrowBtn,
        clientX: 0,
        clientY: 0,
        force: 0,
        radiusX: 0,
        radiusY: 0,
        rotationAngle: 0,
        screenX: 0,
        screenY: 0
      }],
      bubbles: true,
      cancelable: true,
      view: window
    });
    leftArrowBtn.offEvent('touchend', leftArrowBtn);
    leftArrowBtn.dispatchEvent(event2);

    const rightArrowBtn = idsSwapList.shadowRoot.querySelector('.right-arrow');
    const event3 = new TouchEvent('touchend', {
      touches: [{
        identifier: 123,
        pageX: 0,
        pageY: 0,
        target: rightArrowBtn,
        clientX: 0,
        clientY: 0,
        force: 0,
        radiusX: 0,
        radiusY: 0,
        rotationAngle: 0,
        screenX: 0,
        screenY: 0
      }],
      bubbles: true,
      cancelable: true,
      view: window
    });
    rightArrowBtn.offEvent('touchend', rightArrowBtn);
    rightArrowBtn.dispatchEvent(event3);

    setTimeout(() => {
      expect(listItem.selected).toBe('selected');
      expect(idsSwappable.length).toBe(1);
    }, 20);
  });
});
