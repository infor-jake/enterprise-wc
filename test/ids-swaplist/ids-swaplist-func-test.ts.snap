// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsSwapList Component renders correctly 1`] = `"<ids-swaplist id="swaplist-1"></ids-swaplist>"`;
